coffeecard/CoffeeCard.Library/Services/v2/LeaderboardService.cs

```diff
@@ -1,37 +1,92 @@
         }
 
-        public Task<List<LeaderboardEntry>> GetLeaderboard(LeaderboardPreset preset, int top)
+        public async Task<IEnumerable<LeaderboardEntry>> GetTopLeaderboardEntries(LeaderboardPreset preset, int top)
         {
-            var statistics = _context.Statistics
+            var allStatisticsByPreset = await _context.Statistics
                 .Include(s => s.User)
                 .Where(s => s.Preset == preset.ToStatisticPreset())
+                .ToListAsync();
+
+            var topStatistics = allStatisticsByPreset
+                .Where(s => ParsePresetAndCheckValidity(s, preset))
                 .OrderByDescending(s => s.SwipeCount)
                 .ThenBy(s => s.LastSwipe)
                 .Take(top);
 
-            return statistics.Select(s => new LeaderboardEntry
+            return topStatistics.Select((s, index) => new LeaderboardEntry
             {
                 Id = s.User.Id,
                 Name = s.User.PrivacyActivated ? "Anonymous" : s.User.Name,
+                Rank = index + 1,
                 Score = s.SwipeCount
-            }).ToListAsync();
+            });
+        }
+
+        public async Task<LeaderboardEntry> GetLeaderboardEntry(User user, LeaderboardPreset preset)
+        {
+            var allStatisticsByPreset = await _context.Statistics
+                .Where(s => s.Preset == preset.ToStatisticPreset())
+                .ToListAsync();
+
+            var sortedStatistics = allStatisticsByPreset
+                .Where(s => ParsePresetAndCheckValidity(s, preset))
+                .OrderByDescending(s => s.SwipeCount)
+                .ThenBy(s => s.LastSwipe)
+                .ToList();
+
+            var rank = sortedStatistics
+                .FindIndex(s => s.User.Id == user.Id) + 1;
+            var userStatistic = sortedStatistics
+                .FirstOrDefault(s => s.User.Id == user.Id);
+
+            var swipeCount = 0;
+            if (userStatistic != null)
+            {
+                swipeCount = userStatistic.SwipeCount;
+            }
+
+            return new LeaderboardEntry
+            {
+                Id = user.Id,
+                Name = user.Name,
+                Rank = rank,
+                Score = swipeCount
+            };
+        }
+
+        private bool ParsePresetAndCheckValidity(Statistic s, LeaderboardPreset preset)
+        {
+            var now = _dateTimeProvider.UtcNow();
+            
+            return preset switch
+            {
+                LeaderboardPreset.Total => s.Preset == StatisticPreset.Total,
+                LeaderboardPreset.Semester => s.Preset == StatisticPreset.Semester &&
+                                              SemesterUtils.IsSwipeValidInSemester(s.LastSwipe, now),
+                LeaderboardPreset.Month => s.Preset == StatisticPreset.Monthly &&
+                                           SemesterUtils.IsSwipeValidInMonth(s.LastSwipe, now),
+                _ => true
```

Comment:
Should the default case be `false`, in case we add any more `LeaderboardPreset`s in the future and forget to add the case to this switch?

Reply:
I get your point. While not having the strongest opinion, my thought is that we here look at what presets have a restriction on the validity. This is only the case for Semester and Monthly and not the default case of Total. I have refactored the method a bit to reflect this more clearly